openshift/ansible-service-broker
# Add secret support to the Broker

**Describe what this PR does and why we need it**:
This PR allows a broker admin to create associations between apbs and secrets, so that they can be securely added to an APB.

Changes proposed in this pull request
 - Add `secrets` section to broker config. Will associate certain secrets with certain apbs
 - Read in the keys from specified secrets 
   - Mask parameters provided by a secret on a `/catalog` request
- Add volumeMounts and volumes to apb pod to mount secrets into apb container
- if an an apb needs to consume a secret, run it in the broker's namespace with permissions to execute against other namespaces.
- Refactor return values of ExecuteApb into a wrapper struct that contains relevant context (pod name, namespace, serviceaccount)
- Move service account deletion out of the broker and into the apb actions

This is a super initial version of it, right now it's just a list of direct secret-apb associations. WIP is allowing patterns rather than exact matches.

I have a built version of this broker for testing at docker.io/fabianvf/ansible-service-broker:latest

https://trello.com/c/oPvl3Rdy

## Testing

1. Set up an openshift environment with the broker image `fabianvf/origin-ansible-service-broker:latest` and the dockerhub org `fabianvf`. There is a single APB in that org (`rhscl-postgresql-apb`). 

1. Once your cluster is up and my broker is running, run (from the ansible-service-broker project)
    ```bash
    ./scripts/create_broker_secret.py test ansible-service-broker docker.io/fabianvf/rhscl-postgresql-apb  postgresql_user=fabian postgresql_password=testpassword postgresql_version="9.5" 
    ```
1. This will create a secret named test containing the specified keys, and update the broker config to associate it with my APB. 

1. Now, either wait for the catalog to make a new catalog request (this will update the parameters in the UI) (or trigger it if you know how). 

1. Once the catalog request has been made, go to the openshift UI, select the postgresql-apb app, click Next, select whatever namespace you want.
 
1. Note that the parameters `postgresql_user`, `postgresql_password`, `postgresql_version` should not be visible. If you can still see them, either the catalog has not yet updated, or I'm terrible at my job.  

1. Click next and then create.

1. There should be an APB spawned inside the broker namespace. Postgresql should be deployed into whatever namespace you specified. Verify that postgres is up and running in your target namespace.

1. Now create a bind

1. Open up the secret created by the bind, verify that the contents of the secret match the values you passed into the create_secret script.
